src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceTypeOverrideService.kt

```diff
@@ -0,0 +1,124 @@
+  }
+
+  @Transactional
+  fun changeType(licenceId: Long, newType: LicenceType, reason: String) {
+    val licence = licenceRepository.findById(licenceId).orElseThrow { EntityNotFoundException("$licenceId") }
+    val previousType = licence.typeCode
+
+    if (previousType == newType) {
+      log.info("Licence $licenceId already is of type: $newType")
+      return
+    }
+    log.info("Converting Licence $licenceId from $previousType to $newType")
+
+    getIncorrectDates(newType, licence).takeIf { it.isNotEmpty() }?.let { errors ->
+      throw DetailedValidationException(
+        title = "Incorrect dates for new licence type: $newType",
+        errors = errors.map { it.dateName to it.errorType.name },
+      )
+    }
+
+    val username = SecurityContextHolder.getContext().authentication.name
+    val staffMember = staffRepository.findByUsernameIgnoreCase(username)
+
+    licence.typeCode = newType
+
+    val (relevantConditions, conditionsToRemove) = licence.getAdditionalConditionsByRelevancy()
+    val standardConditions = licencePolicyService.getStandardConditionsForLicence(licence)
+
+    val licenceToPersist = licence.updateConditions(
+      updatedAdditionalConditions = relevantConditions,
+      updatedStandardConditions = standardConditions,
+      staffMember = staffMember,
+    )
+
+    licenceRepository.saveAndFlush(licenceToPersist)
+
+    auditEventRepository.saveAndFlush(
+      AuditEvent(
+        licenceId = licence.id,
+        detail = "ID ${licence.id} type ${licence.typeCode} status ${licence.statusCode} version ${licence.version}",
+        eventTime = LocalDateTime.now(),
+        eventType = AuditEventType.USER_EVENT,
+        username = staffMember?.username ?: SYSTEM_USER,
+        fullName = staffMember?.fullName ?: SYSTEM_USER,
+        summary = "Licence type overridden for ${licence.forename} ${licence.surname}: from $previousType to $newType: $reason",
+        changes = mutableMapOf(
+          "oldType" to previousType.name,
+          "newType" to newType.name,
+          "deletedAdditionalConditions" to conditionsToRemove.map {
+            mapOf(
+              "code" to it.conditionCode,
+              "version" to it.conditionVersion,
+              "category" to it.conditionCategory,
+              "text" to it.expandedConditionText,
+            )
+          },
+        ),
+      ),
+    )
+  }
+
+  private fun Licence.getAdditionalConditionsByRelevancy() =
+    this.additionalConditions.partition { this.typeCode.conditionTypes().contains(it.conditionType) }
+
+  private fun getIncorrectDates(licenceType: LicenceType, licence: Licence): Set<IncorrectDate> {
+    val led = licence.licenceExpiryDate
+    val tused = licence.topupSupervisionExpiryDate
+
+    val errors = mutableSetOf<IncorrectDate>()
+    when (licenceType) {
+      AP -> {
+        if (led == null) errors.add(IncorrectDate("LED", MISSING_DATE))
+        if (tused != null) errors.add(IncorrectDate("TUSED", PRESENT_DATE))
```

Comment:
I'm not sure if having a past TUSED is "valid" (or rather, not invalid)

Reply:
It looks like we do `if !tused or past tused -> AP`
[Here](https://github.com/ministryofjustice/create-and-vary-a-licence-api/blob/555ca05735ff1ca98f14a41c2179b4a8a91e6fb8/src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/util/LicenceType.kt#L16)